src/sitemap/handler.js

```diff
@@ -10,131 +10,309 @@
+      }
+    }
+  });
+
+  // Further process matching URLs if necessary
+  const response = {};
+  const pagesPromises = matchingUrls.map(async (matchingUrl) => {
+    // Check if further detailed checks are needed or directly use cached data
+    if (!contentsCache[matchingUrl]) {
+      contentsCache[matchingUrl] = await checkSitemap(matchingUrl);
     }
+    const pages = await getBaseUrlPagesFromSitemapContents(
+      baseUrl,
+      contentsCache[matchingUrl].details,
+    );
 
-    const baseURL = site.getBaseURL();
-    const auditResult = await findSitemap(baseURL);
+    if (pages.length > 0) {
+      response[matchingUrl] = pages;
+    }
+  });
 
-    log.info(`Audit result for ${baseURL}:\n${JSON.stringify(auditResult, null, 2)}`);
+  // Wait for all pages promises to resolve
+  await Promise.all(pagesPromises);
 
-    await sqs.sendMessage(queueUrl, {
-      type,
-      url: baseURL,
-      auditContext,
-      auditResult,
+  return response;
+}
+/**
+ * This function is used to find the sitemap of a given URL.
+ * It first extracts the domain and protocol from the input URL.
+ * If the URL is invalid, it returns an error message.
+ * If the URL is valid, it checks the sitemap path in the robots.txt file.
+ * Then toggles the input URL with www and filters the sitemap URLs.
+ * It then gets the base URL pages from the sitemaps.
+ * The extracted paths response length > 0, it returns the success status, log messages, and paths.
+ * The extracted paths response length < 0, log messages and returns the failure status and reasons.
+ *
+ * @param {string} inputUrl - The URL for which to find and validate the sitemap
+ * @param {UniversalContext} context - The Lambda context object
+ * @returns {Promise<{success: boolean, reasons: Array<{value}>, paths?: any}>} result of sitemap
+ */
+
+export async function findSitemap(inputUrl, context) {
+  const logMessages = [];
+
+  const parsedUrl = extractDomainAndProtocol(inputUrl);
+  if (!parsedUrl) {
+    logMessages.push({
+      value: inputUrl,
+      error: ERROR_CODES.INVALID_URL,
     });
+    return {
+      success: false,
+      reasons: logMessages,
+    };
+  }
 
-    log.info(`Successfully audited ${baseURL} for ${type} type audit`);
+  const { protocol, domain } = parsedUrl;
+  let sitemapUrls = [];
+  try {
+    const robotsResult = await checkRobotsForSitemap(protocol, domain);
+    if (robotsResult.paths.length) {
+      sitemapUrls = robotsResult.paths;
+    }
+  } catch (error) {
+    logMessages.push({ value: `Error fetching or processing robots.txt: ${error.message}`, error: ERROR_CODES.FETCH_ERROR });
+    // Don't return failure yet, try the fallback URLs
+  }
 
-    return noContent();
-  } catch (e) {
-    return internalServerError('Sitemap audit failed');
+  if (!sitemapUrls.length) {
+    const commonSitemapUrls = [`${protocol}://${domain}/sitemap.xml`, `${protocol}://${domain}/sitemap_index.xml`];
+    sitemapUrls = await checkCommonSitemapUrls(commonSitemapUrls);
```

Comment:
makes sense. Also as an idea, we could try to get the (maybe hidden) sitemaps from Google Search Console, when the customer opted in for that. I can add another function in the Google Client for that.

Reply:
> businesses sometimes hide their sitemaps, often for valid reasons. I think if we cannot locate the sitemap, it shouldn't be flagged as an error. rather, we should just skip the site for sitemap audit altogether

@ekremney This is debatable :) since the existence of a sitemap is important for indexing and managing the crawling budget. Instead of an error, it could be marked as a warning or recommendation (despite SEO tools like Ahrefs flagging it as an error). There are small websites (around 100 pages) that might not require a sitemap, but it is still recommended from an SEO standpoint.

Reply:
> makes sense. Also as an idea, we could try to get the (maybe hidden) sitemaps from Google Search Console, when the customer opted in for that. I can add another function in the Google Client for that.

@dzehnder This could be an option to expand the detection area but it relies on the customer to add/remove the sitemap(s) in GSC. Some sitemaps may not be submitted in GSC, but can be seen in /robots.txt file

Reply:
> since the existence of a sitemap is important for indexing and managing the crawling budget. Instead of an error, it could be marked as a warning or recommendation (despite SEO tools like Ahrefs flagging it as an error). There are small websites (around 100 pages) that might not require a sitemap, but it is still recommended from an SEO standpoint.

they hide their content to prevent competitors from crawling their content. Hiding sitemap is not necessarily a bad thing, GSC manually: https://developers.google.com/search/docs/crawling-indexing/sitemaps/build-sitemap#addsitemap instead of advertising in robots.txt